hibernate-core/src/main/java/org/hibernate/query/criteria/internal/predicate/InPredicate.java

```diff
@@ -197,14 +201,51 @@ public String render(boolean isNegated, RenderingContext renderingContext) {
 				}
 			}
 			else {
+				Expression lastValue = null;
 				buffer.append( '(' );
 				String sep = "";
 				for ( Expression value : values) {
+					lastValue = value;
 					buffer.append( sep )
 							.append( ( (Renderable) value )
 									.render( renderingContext ) );
 					sep = ", ";
 				}
+
+
+				if (renderingContext.getCriteriaLiteralHandlingMode() == LiteralHandlingMode.BIND &&
+						lastValue instanceof LiteralExpression) {
+
+					int bindValueCount = values.size();
+					int bindValueMaxCount = bindValueCount;
+
+					final SessionFactoryImplementor sfi = criteriaBuilder().getEntityManagerFactory().unwrap( SessionFactoryImplementor.class );
+					final Dialect dialect = sfi.getServiceRegistry().getService(JdbcServices.class).getJdbcEnvironment().getDialect();
+					final int inExprLimit = dialect.getInExpressionCountLimit();
+
+					boolean inClauseParameterPaddingEnabled = sfi.getSessionFactoryOptions().inClauseParameterPaddingEnabled() &&
+									bindValueCount > 2;
+
+					if ( inClauseParameterPaddingEnabled ) {
+						int bindValuePaddingCount = MathHelper.ceilingPowerOfTwo( bindValueCount );
+
+						if (inExprLimit > 0 && bindValuePaddingCount > inExprLimit ) {
+							bindValuePaddingCount = Math.max(bindValueCount, inExprLimit);
+						}
+
+						if ( bindValueCount < bindValuePaddingCount ) {
+							bindValueMaxCount = bindValuePaddingCount;
+						}
+
+						for (int i = bindValueCount; i < bindValueMaxCount; i++) {
+							buffer.append( sep )
+									.append( ( (Renderable) lastValue )
+											.render( renderingContext ) );
```

Comment:
Do you have some formatting issue above for as you can see the method chaining are not aligned vertically correctly.

Reply:
I just copied lines 209-211. I thought the formatting is ok, as it was done by a member of the hibernate team.